sql/2021/pwa/top_manifest_categories.sql

```diff
@@ -0,0 +1,89 @@
+#standardSQL
+# Top manifest categories
+CREATE TEMPORARY FUNCTION getCategories(manifest STRING)
+RETURNS ARRAY<STRING> LANGUAGE js AS '''
+try {
+  var $ = Object.values(JSON.parse(manifest))[0];
+  var categories = $.categories;
+  if (typeof categories == 'string') {
+    return [categories];
+  }
+  return categories;
+} catch (e) {
+  return null;
+}
+''';
+SELECT
+  'PWA Sites' AS type,
+  _TABLE_SUFFIX AS client,
+  category,
+  COUNT(0) AS freq,
+  total,
+  COUNT(0) / total AS pct
+FROM
+  `httparchive.pages.2021_07_01_*`,
+  UNNEST(getCategories(JSON_EXTRACT(payload, '$._pwa.manifests'))) AS category
+JOIN
+  (
+    SELECT
+      _TABLE_SUFFIX,
+      COUNT(0) AS total
+    FROM
+      `httparchive.pages.2021_07_01_*`
+    WHERE
+      JSON_EXTRACT(payload, '$._pwa') != "[]" AND
+      JSON_EXTRACT(payload, '$._pwa.manifests') != "[]" AND
+      JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = "true"
```

Comment:
There are 12 `JSON_EXTRACT` statements in this query. Any way to simplify that?

Reply:
To be honest there's not real reason for the first one (`JSON_EXTRACT(payload, '$._pwa') != "[]"`) since the second and third covers this. Will remove from all queries which will bring it down a bit. But still be a lot of these.

Reply:
As per above idea, this (and all the other queries run for both All and PWA Sites) could be refactored to do the main select in a CTE like this, but honestly think the `UNION ALL` is more easily understood. Last year these were separate queries but this year I decided to merge precisely because there was a lot of repetition.

```sql
#standardSQL
# Top manifest categories
CREATE TEMPORARY FUNCTION getCategories(manifest STRING)
RETURNS ARRAY<STRING> LANGUAGE js AS '''
try {
  var $ = Object.values(JSON.parse(manifest))[0];
  var categories = $.categories;
  if (typeof categories == 'string') {
    return [categories];
  }
  return categories;
} catch (e) {
  return null;
}
''';

WITH totals AS (
  SELECT
    _TABLE_SUFFIX,
    COUNT(0) AS total,
    COUNTIF(JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = "true") AS pwa_total
  FROM
    `httparchive.pages.2021_07_01_*`
  WHERE
    JSON_EXTRACT(payload, '$._pwa.manifests') != "[]"
  GROUP BY
    _TABLE_SUFFIX
),

manifests_categories AS (
  SELECT
    'All Sites' AS type,
    _TABLE_SUFFIX AS client,
    category,
    COUNT(0) AS freq,
    total,
    COUNT(0) / total AS pct,
    COUNTIF(JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = "true") AS pwa_freq,
    pwa_total,
    COUNTIF(JSON_EXTRACT(payload, '$._pwa.serviceWorkerHeuristic') = "true") / pwa_total AS pwa_pct
  FROM
    `httparchive.pages.2021_07_01_*`,
    UNNEST(getCategories(JSON_EXTRACT(payload, '$._pwa.manifests'))) AS category
  JOIN
    totals
  USING (_TABLE_SUFFIX)
  WHERE
    JSON_EXTRACT(payload, '$._pwa.manifests') != "[]"
  GROUP BY
    client,
    category,
    total,
    pwa_total
  HAVING
    category IS NOT NULL
  ORDER BY
    type DESC,
    freq / total DESC,
    category,
    client
)

SELECT
  'PWA Sites' AS type,
  client,
  category,
  pwa_freq AS freq,
  pwa_total AS total,
  pwa_pct AS pct
FROM
  manifests_categories
UNION ALL
SELECT
  'All Sites' AS type,
  client,
  category,
  freq,
  total,
  pct
FROM
  manifests_categories
ORDER BY
  type DESC,
  pct DESC,
  category,
  client
```

Reply:
Decided to go with the CTE method for these more complex queries. Have left the simpler ones as is.